### PR TITLE
chore(lib): refactored lib utils and groups

### PR DESCRIPTION
**Pull request by pranshuchittora**
**What kind of change does this PR introduce?**
     Refactoring
**Did you add tests for your changes?**
    No
**If relevant, did you update the documentation?**
    No
**Summary**
- Object destructing to prevent deep level object refs (quite confusing and tedious to write)
- repeated else if to switch

**Does this PR introduce a breaking change?**
No

**Other information**

### CODE REVIEW

File: lib/utils/interactive.js [hunk 1]
```diff
@@ -124,12 +124,19 @@ module.exports = async function(config, outputOptions, processingMessageBuffer)
...
+                process.stdout.write(ansiEscapes.cursorNextLine);
+                break;
+            case 'up':
+                process.stdout.write(ansiEscapes.cursorPrevLine);
+                break;
+            case 'return':
+                // TODO: get line and do stuff
+                break;
+            default:
+                break;
```
Comment (rishabh3112) [inline lib/utils/interactive.js]: It's my personal opinion, I think `if else` provides more flexibility for future need as compared to `switch` and are comparatively more readable too. Consider if you like.

File: lib/groups/basic.js [hunk 2]
```diff
@@ -8,24 +8,32 @@ class BasicGroup extends GroupHelper {
...
+                case 'sourcemap':
+                    this.opts.options.devtool = args[arg] || 'eval';
+                    outputOptions.devtool = args[arg];
+                    break;
+
+                case 'entry':
+                    this.opts.options[arg] = this.resolveFilePath(args[arg], 'index.js');
+                    break;
+                default:
+                    break;
```
Comment (rishabh3112) [inline lib/groups/basic.js]: Same here

File: lib/utils/interactive.js (hunk 1 above)
Comment (pranshuchittora) [reply lib/utils/interactive.js]: I completely agree that `if-else` are easier to read.

> As it turns out, the switch statement is faster in most cases when compared to if-else, but significantly faster only when the number of conditions is large. The primary difference in performance between the two is that the incremental cost of an additional condition is larger for if-else than it is for switch.

https://www.oreilly.com/library/view/high-performance-javascript/9781449382308/ch04.html#:~:targetText=As%20it%20turns%20out%2C%20the,than%20it%20is%20for%20switch%20.

Comment (rishabh3112) [reply lib/utils/interactive.js]: @pranshuchittora a valid point. My personal opinions ahead:
That excerpt reminded me of my debate with 1st semester C language teacher why one would use switch over if else and this was the major point for me to get satisfied. Still, If a particular block of code is to be executed only one, I would prefer `if-else` only as the gain is not that much.

Comment (pranshuchittora) [reply lib/utils/interactive.js]: :smile: That's a good old custom in the engineering of having debates with the profs. Switch case won't match for all conditions if `break` is there.

Comment (rishabh3112) [reply lib/utils/interactive.js]: Ofcourse, I meant to say this:
execution of whole switch if we writing switch in a for loop, thats the adequate performance gain.
But if the conditions are meant checked only for a single time, then we need not have a switch.

Comment (pranshuchittora) [reply lib/utils/interactive.js]: Uhm.. But I think a switch is an optimal & feasible choice.

File: lib/utils/compiler.js [hunk 3]
```diff
@@ -41,19 +41,22 @@ function setUpHookForCompiler(compiler, outputOptions, options) {
     });
 
     if (outputOptions.infoVerbosity === 'verbose') {
+        const resolveCompilationName = compilation => {
```
Comment (jamesgeorge007) [inline lib/utils/compiler.js]: Why not make use of the inline approach?

```js
        const resolveCompilationName = compilation => 
          compilation.name ? compilation.name : '';
```

Comment (pranshuchittora) [reply lib/utils/compiler.js]: :thinking:  IMO this is much more inferable.

Comment (jamesgeorge007) [reply lib/utils/compiler.js]: Just personal preference :smile:

File: lib/utils/compiler.js [hunk 4]
```diff
@@ -41,19 +41,22 @@ function setUpHookForCompiler(compiler, outputOptions, options) {
...
+            return compilation.name ? compilation.name : '';
+        };
         if (outputOptions.watch) {
             compiler.hooks.watchRun.tap('WebpackInfo', compilation => {
-                const compilationName = compilation.name ? compilation.name : '';
+                const compilationName = resolveCompilationName(compilation);
```
Comment (ematipico) [inline lib/utils/compiler.js]: You could actually remove the need of a variable:
```js
process.cliLogger.info('Compilation ' + resolveCompilationName(compilation) + ' starting…'); 
```

File: lib/groups/advanced.js [hunk 5]
```diff
@@ -43,33 +43,34 @@ class AdvancedGroup extends GroupHelper {
     }
     resolveOptions() {
         const { args } = this;
+        const Options = this.opts.options;
```
Comment (ematipico) [inline lib/groups/advanced.js]: Lower case `options`

File: lib/utils/zero-config.js [hunk 6]
```diff
@@ -1,8 +1,8 @@
 function getEnvFromOptionsAndMode(mode, optionsObject) {
     const { dev, prod } = optionsObject;
-
-    if ((process.env.NODE_ENV && process.env.NODE_ENV === 'production') || process.env.NODE_ENV === 'development') {
-        return process.env.NODE_ENV;
+    const NODE_ENV = process.env.NODE_ENV;
+    if ((NODE_ENV && NODE_ENV === 'production') || NODE_ENV === 'development') {
```
Comment (rishabh3112) [inline lib/utils/zero-config.js]:
```suggestion
    if (NODE_ENV && (NODE_ENV === 'production' || NODE_ENV === 'development')) {
```

File: lib/groups/AdvancedGroup.js [hunk 7]
```diff
@@ -44,33 +44,34 @@ class AdvancedGroup extends GroupHelper {
...
-            if (this.opts.options && this.opts.options.plugins) {
-                this.opts.options.plugins.unshift(prefetchVal);
+            if (options && options.plugins) {
+                options.plugins.unshift(prefetchVal);
             } else {
-                this.opts.options.plugins = [prefetchVal];
+                options.plugins = [prefetchVal];
             }
         }
         if (args.plugin) {
```
Comment (ematipico) [inline lib/groups/AdvancedGroup.js]: Please comment this check. Plugin function is commented at the moment.

File: lib/groups/AdvancedGroup.js [hunk 8]
```diff
@@ -44,33 +44,34 @@ class AdvancedGroup extends GroupHelper {
     // }
     resolveOptions() {
         const { args } = this;
+        const options = this.opts.options;
```
Comment (jamesgeorge007) [inline lib/groups/AdvancedGroup.js]:
```suggestion
        const { options } = this.opts;
```

File: lib/utils/zero-config.js [hunk 9]
```diff
@@ -1,8 +1,8 @@
 function getEnvFromOptionsAndMode(mode, optionsObject) {
     const { dev, prod } = optionsObject;
```
Comment (jamesgeorge007) [inline lib/utils/zero-config.js]:
```js
function getEnvFromOptionsAndMode(mode, {dev, prod}) {
```

Comment (ematipico) [reply lib/utils/zero-config.js]: That's not my favourite suggestion. I would keep it like this. Destructuring is fine when there's only one single object passed to the function. Here, we don't know what's inside the second parameter (variable names help)

File: lib/groups/AdvancedGroup.js (hunk 8 above)
Comment (ematipico) [reply lib/groups/AdvancedGroup.js]: At this point you should suggest:

```js
const { args, opts: { options } } = this;
```

:)

Comment (jamesgeorge007) [reply lib/groups/AdvancedGroup.js]: Deep nesting might affect code readability.

File: lib/utils/zero-config.js (hunk 9 above)
Comment (jamesgeorge007) [reply lib/utils/zero-config.js]: Yeah, makes sense.

**Comment by ematipico** [comment]
Thank you!

**Pull request merged**